src/ServiceControl.Monitoring/QueueLength/QueueLengthDataStore.cs

```diff
@@ -1,117 +1,103 @@
         }
 
-        public Dictionary<string, Dictionary<DateTime, double>> GetQueueLengths(DateTime now)
+        public void FillIn(MonitoredEndpoint[] data, DateTime now)
         {
-            var results = new Dictionary<string, Dictionary<DateTime, double>>();
+            var snapshot = GetIntervals(now).ToLookup(k => $"{k.Id.EndpointName}");
 
-            foreach (var queueLength in queueLengths)
+            foreach (var item in data)
             {
-                var endpointName = queueLength.Key;
-
-                ConcurrentDictionary<DateTime, double> endpointData;
-
-                if (queueLengths.TryGetValue(endpointName, out endpointData))
-                {
-                    var endpointDataSnapshot = endpointData.ToArray();
-
-                    var points = endpointDataSnapshot
-                        .Where(kvp => kvp.Key > now.Subtract(HistoricalPeriod))
-                        .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
-
-                    results.Add(endpointName, points);
-
-                    var outdatedPoints = endpointData.Where(kvp => kvp.Key <= now.Subtract(HistoricalPeriod));
-
-                    foreach (var outdatedPoint in outdatedPoints)
-                    {
-                        double removedValue;
-                        endpointData.TryRemove(outdatedPoint.Key, out removedValue);
-                    }
-                }
+                item.QueueLength = AggregateQueueLength(snapshot[item.Name].ToList());
             }
-
-            return results;
         }
 
         public void Store(EndpointInstanceId endpointId, JObject data)
         {
-            var updatedEndpoints = new List<string>();
-
             var counters = (JArray) data["Counters"] ?? new JArray();
-            updatedEndpoints.AddRange(UpdateSends(counters));
+            UpdateSends(counters);
 
             var gauges = (JArray) data["Gauges"] ?? new JArray();
-            updatedEndpoints.AddRange(UpdateReceives(endpointId.EndpointName, gauges));
+            UpdateReceives(endpointId.EndpointName, gauges);
+        }
 
+        public void SnapshotCurrentQueueLengthEstimations(DateTime now)
+        {
             var virtualQueueLengths = calculator.GetQueueLengths();
 
-            foreach (var updatedEndpoint in updatedEndpoints.Distinct())
+            var queueLengths = virtualQueueLengths
+                .GroupBy(vq => vq.Key.EndpointName)
+                .Select(g => new
+                {
+                    Id = new EndpointInstanceId(g.Key, string.Empty),
+                    Entry = new RawMessage.Entry
+                    {
+                        DateTicks = now.Ticks,
+                        Value = (long)g.Sum(i => i.Value)
+                    }
+                });
+
+            foreach (var queueLength in queueLengths)
             {
-                SnapshotValueForEndpoint(virtualQueueLengths, updatedEndpoint);
+                Store(queueLength.Id, new[]
```

Comment:
The `IntervalsStore` works in following way. If there is an entry responsible for storing data within the given period it will add the number [to the specific bucket](https://github.com/Particular/ServiceControl.Monitoring/blob/c373b51c8cf760098b413f4ff4c19f00f6d06ff3/src/ServiceControl.Monitoring/Timings/IntervalsStore.cs#L119).

```
intervals[epochIndex].TotalTime += entry.Value;
intervals[epochIndex].TotalMeasurements += 1;
```

Consider a following scenario: the queue length does not change and is reported second after second which fits into one bucket. Then the `Value` will be bumped up multiple times with the same value.

Am I seeing it wrong?

We could check this by snapshotting the value twice in one test and see if it's different.